src/transformers.jl

```diff
@@ -28,30 +28,80 @@ Non-mutating version of [`transform!`](@ref).
 function transform end
 
 """
-    Transforms.apply!(data::T, ::Transform; kwargs...) -> T
+    apply(data::T, ::Transform; kwargs...)
+
+Applies the [`Transform`](@ref) to the data. New transforms should usually only extend
+`_apply` which this method delegates to.
+
+Where possible, this should be extended for new data types `T`.
+"""
+function apply end
+
+"""
+    apply!(data::T, ::Transform; kwargs...) -> T
+
+Applies the [`Transform`](@ref) mutating the input `data`. New transforms should usually
+only extend `_apply!` which this method delegates to.
 
-Applies the [`Transform`](@ref) mutating the input `data`.
 Where possible, this should be extended for new data types `T`.
 """
 function apply! end
 
+
 """
-    Transforms.apply(data::T, ::Transform; kwargs...) -> T
+    apply(A::AbstractArray, ::Transform; dims=:, inds=:, kwargs...)
 
-Non-mutating version of [`apply!`](@ref), which it delegates to by default.
-Does not need to be extended unless a mutating [`Transform`](@ref) is not possible.
+Applies the [`Transform`](@ref) to each element of `A`.
+Optionally specify the `dims` to apply the [`Transform`](@ref) along certain dimensions
+and `inds` will be the indices to apply the Transform to along the `dims` specified.
+If `dims` === : (all dimensions), then `inds` will be the global indices of the array,
+instead of being relative to a certain dimension.
+
+Will return the same datatype except for AxisArrays as operations on those do not preserve
+type.
 """
-function apply end
+function apply(A::AbstractArray, t::Transform; dims=:, inds=:, kwargs...)
+    if dims === Colon()
+        if inds === Colon()
+            return _apply(A, t; kwargs...)
+        else
+            return _apply(A[:][inds], t; kwargs...)
+        end
+    end
+
+    return mapslices(x -> _apply(x[inds], t; kwargs...), A, dims=dims)
+end
 
 """
-    apply!(A::AbstractArray{T}, ::Transform; dims=:, kwargs...) where T <: Real
+    apply(table, ::Transform; cols=nothing, kwargs...)
+
+Applies the [`Transform`](@ref) to each of the specified columns in the `table`.
+If no `cols` are specified, then the [`Transform`](@ref) is applied to all columns.
+
+Returns an array containing each transformed column, in the same order as `cols`.
+"""
+function apply(table, t::Transform; cols=nothing, kwargs...)
+    Tables.istable(table) || throw(MethodError(apply!, (table, t)))
+
+    # Extract a columns iterator that we should be able to use to mutate the data.
+    # NOTE: Mutation is not guaranteed for all table types, but it avoid copying the data
+    columntable = Tables.columns(table)
+
+    cnames = cols === nothing ? propertynames(columntable) : cols
+    return [_apply(getproperty(columntable, cname), t; kwargs...)  for cname in cnames]
+end
+
+#TODO: should this be apply! ?
```

Comment:
I'm not sure what the difference is from calling `apply!` here instead of `_apply!`. 

The general `apply!` still calls `_apply!` so I think it just performs unneeded checks again? I may be wrong

Reply:
Yeah, I think it should be `_apply!` rather than `apply!`.

However, I'm thinking about the move towards `apply` being more standard than `apply!`. The way things are, `_apply!` must be implemented for each transform.

If we want to switch things around, we could have a default `_apply!` similar to what you added in `periodic.jl`:

```
function _apply!(x, t::Transform; kwargs...)
    x[:] = _apply(x, t; kwargs...)
    return x
end
```

Not essential for this PR though.

Reply:
Maybe the error if `_apply!` is not implemented is more clear though? Because not all transforms support non-mutating, and the above example will result in Conversion errors. Is a method not implemented error more clear than a conversion error?

Reply:
Hmm I see where you're coming from. Method not implemented error does seem clearer - "we don't support mutating for this transform" rather than "we tried and failed to support mutating for this transform".

Reply:
Yeah I think it should be:
```
_apply(x, t::Transform; kwargs...) = _apply!(_try_copy(x), t; kwargs...)
```

The thinking was that you would only have to implement the mutating `_apply!` version for each new Transform.
Then for the non-mutating call to `_apply`, it should just copy the input and call `_apply!`.

So this was supposed to be a convenience for not having to define `_apply` for every new transform.

>The way things are, _apply! must be implemented for each transform.

yeah our philosophy has since changed to "Every transform must implement `_apply` and, where possible, `_apply!`, wherein `_apply` can just call that after copying the data".

So I think this  can still be suported?